### PR TITLE
[WIP] feat(volume resize, mayactl, cstor): support for cstor volume resize via mayactl

### PR DESCRIPTION
**Pull request by mittachaitu**
### What this PR does ?
This PR implements support for cstor volume resize via mayactl command.

### Prerequisites for cstor volume resize:
- User needs to back up data for those volumes that should be resize.
- User needs to make sure that application pod shouldn't have any active IOs.

### Steps to resize cstor volumes
- Exec into maya-apiserver pod and execute below command:
  - mayactl volume resize --volname <volume_name> --size <size><unit> --namespace \<namespace>
  - where namespace is pvc's namespace.
----------------------------------------------------------------------------------------------------------------------------------
**Example**:
1)   maya-apiserver pod can be get by executing below command
```
kubectl get pods -n openebs
NAME                                                              READY   STATUS    RESTARTS   AGE
cstor-sparse-pool-asn1-744b757bb9-ndtlk                           2/2     Running   0          68m
maya-apiserver-76848cc4b-x4ccc                                    1/1     Running   0          69m
openebs-admission-server-589d7b7dc-nm27c                          1/1     Running   0          69m
openebs-ndm-bh55k                                                 1/1     Running   0          69m
openebs-provisioner-c44b4fbfc-9nzjh                               1/1     Running   0          69m
openebs-snapshot-operator-65c946b776-jdnj7                        2/2     Running   0          69m
pvc-c6a04572-498b-11e9-a465-28d2440cdb59-target-6dc8767b757gbbz   3/3     Running   0          66s
```
2) Exec into the maya-apiserver 
```   kubectl exec -it <maya-apiserver-76848cc4b-x4ccc> -n openebs bash```
3) Execute the resize command inside maya-apiserver
```mayactl volume resize --volname <volume_name> --size 6G -n <namespace>```
 In case of success output of the above command will be:
```
Executing the volume resize...
Volume resize successfull for volume: '<volume_name>: '6G' in '<namespace_name>' namespace
```
----------------------------------------------------------------------------------------------------------------------------------
- Once it is successful then execute below post resize steps:
-  Execute below command on application node to rescan the iscsi session.
  - `sudo iscsiadm -m node -R`
- Extend the file system used by application pod on a scheduled application node
  - to know the disk used by the application execute below commands in app node
  - `sudo iscsiadm -m node -p 3`
  - extend the filesystem: `resize2fs /dev/sd*`

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes https://github.com/openebs/openebs/issues/1047

**TODO**:
- Event generation in cstor-volume-management at the time of re-writing istgt.conf file.
- Updating the pvc status field with resized capacity.
 

**Special notes for your reviewer**:
Current resize operation will not support updating ```PVC``` which will be addressed by ```OpenEBS CSI driver``` implementation.

### CODE REVIEW

**Comment by AmitKumarDas** [comment]
@mittachaitu can you put more concrete info on following:
```
**How to get cStor volume resize**:

* Exec into maya-apiserver pod and execute below command
  mayactl volume resize --volname <volume_name> --size  --namespace 
  Once it is successfull then as post resize steps:-
* Bring the application into running state.
* Extend the file system of application on the application node.
```

File: cmd/mayactl/app/command/volume.go [hunk 1]
```diff
@@ -95,6 +95,13 @@ Examples:
 
  # Delete a Volume created in 'test' namespace:
    $ mayactl volume delete --volname <vol> --namespace test
+
+ # Resize a Volume created in default namespace:
+   $ mayactl volume resize --volname <vol> --size <size><unit>
```
Comment (AmitKumarDas) [inline cmd/mayactl/app/command/volume.go]: Can you comment on supported units

File: cmd/mayactl/app/command/volume.go [hunk 2]
```diff
@@ -95,6 +95,13 @@ Examples:
...
+
+ # Resize a Volume created in 'test' namespace:
+   $ mayactl volume resize --volname <vol> --size <size><unit> --namespace test
```
Comment (AmitKumarDas) [inline cmd/mayactl/app/command/volume.go]: what are the supported units?

File: cmd/mayactl/app/command/volume_resize.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	cmd.Flags().StringVarP(&options.size, "size", "", options.size,
+		"expanding size")
+	cmd.MarkPersistentFlagRequired("size")
+	return cmd
+}
+
+// ValidateResize validates the flag values
+func (c *CmdVolumeOptions) ValidateResize(cmd *cobra.Command) error {
+	if len(c.volName) == 0 {
+		return errors.New("--volname is missing. Please specify an unique name")
```
Comment (AmitKumarDas) [inline cmd/mayactl/app/command/volume_resize.go]: Please specify the name of the volume to be resized

File: cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go [hunk 4]
```diff
@@ -70,6 +71,9 @@ func (s *HTTPServer) volumeV1alpha1SpecificRequest(resp http.ResponseWriter, req
 		cvol, err := volOp.httpDelete()
 		sendEventOrIgnore(cvol, usage.VolumeDeprovision)
 		return cvol, err
+	case "UPDATE":
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go]: Added PUT method

File: pkg/client/volume/cstor/v1alpha1/cstor.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package v1alpha1
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/openebs/maya/pkg/client/generated/cstor-volume-mgmt/v1alpha1"
```
Comment (AmitKumarDas) [inline pkg/client/volume/cstor/v1alpha1/cstor.go]: can you alias this to "cstor_vol_mgmt_client"

File: pkg/install/v1alpha1/cstor_volume.go [hunk 6]
```diff
@@ -1003,7 +1036,139 @@ spec:
...
+metadata:
+  name: cstor-volume-resize-resizevolume-default
+spec:
+  meta: |
+    id: resizecstorvolume
+    kind: Command
+  post: |
+    {{- $runCommand := update cstor volume | withoption "ip" .TaskResult.resizelistsvc.clusterIP -}}
+    {{- $runCommand := $runCommand | withoption "volname" .Volume.owner -}}
+    {{- $runCommand | withoption "capacity" .Volume.capacity | run | saveas "resizecstorvolume" .TaskResult -}}
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/install/v1alpha1/cstor_volume.go]: ```resize cstor volume``` will good. I will update it with ```resize cstor volume```

File: pkg/validation/string_validation.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// ValidateString checks whether the string matches with provided regular
+// expression or not
+func ValidateString(str, expr string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/validation/string_validation.go]: @AmitKumarDas we are required validation of size in mayactl as well as maya-apiserver side. So I made a separate function for evaluating regex.

File: pkg/volume/cstor/v1alpha1/cstor.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+// cstor is used to invoke resize call
+type cstor struct {
+	IP  string
+	Vol *v1alpha1.CASVolume
+}
+
+// Cstor return a pointer to cstor
+func Cstor() *cstor {
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/volume/cstor/v1alpha1/cstor.go]: Implemented above builder pattern.

File: pkg/task/task.go [hunk 9]
```diff
@@ -644,6 +646,50 @@ func (m *taskExecutor) patchOEV1alpha1CVR() (err error) {
...
+	}
+	pv, err := m.getK8sClient().PatchCoreV1PVAsRaw(m.getTaskObjectName(), pe.patchType(), raw)
+	if err != nil {
+		return
+	}
+	util.SetNestedField(m.templateValues, pv, string(v1alpha1.CurrentRepeatResourceLITP))
+	return
+}
+
+// patchCoreV1PVC will patch persistentvolume claim as defined in the runtask
```
Comment (AmitKumarDas) [inline pkg/task/task.go]: remove this code

File: pkg/string/v1alpha1/string.go [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	lib_strings "strings"
+)
+
+type list struct {
+	items []string
+}
+
+// List will make list of sstring slices
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/string/v1alpha1/string.go]: I will update the comment.

File: cmd/mayactl/app/command/volume.go (hunk 1 above)
Comment (prateekpandey14) [reply cmd/mayactl/app/command/volume.go]: @mittachaitu let's add one resize test using `mayactl volume resize ` for any Cstor volume we have in travis tests

File: cmd/mayactl/app/command/volume.go [hunk 11]
```diff
@@ -95,6 +95,14 @@ Examples:
 
  # Delete a Volume created in 'test' namespace:
    $ mayactl volume delete --volname <vol> --namespace test
+
+ # Resize a Volume created in default namespace:
+   $ mayactl volume resize --volname <vol> --size <size><unit>
```
Comment (prateekpandey14) [inline cmd/mayactl/app/command/volume.go]: @mittachaitu let's add one resize test using `mayactl volume resize` for any Cstor volume we have in travis tests

Comment (mittachaitu) [reply cmd/mayactl/app/command/volume.go]: @prateekpandey14 istgt doesn't support resize, to support resize operation at istgt(https://github.com/mittachaitu/istgt/commit/25c02c152159336577d786460506916b59fc53c6) working on the UT to raise a PR.

File: cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go [hunk 12]
```diff
@@ -91,6 +95,16 @@ func (v *volumeAPIOpsV1alpha1) httpGet() (interface{}, error) {
 	return v.read(volName)
 }
 
+//httpPut deals with http PUT request
+func (v *volumeAPIOpsV1alpha1) httpPut() (interface{}, error) {
+	path := strings.TrimSpace(strings.TrimPrefix(v.req.URL.Path, "/latest/volumes"))
+	if strings.Contains(path, "/resize/") {
+		return v.resize()
+	} else {
```
Comment (muratkars) [inline cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go]: No need for else here. The usual Go style for that is

if strings.Contains(path, "/resize/") {
		return v.resize()
	}
	return nil, CodedError(405, fmt.Sprintf("failed to process a PUT request"))
	}

File: pkg/install/v1alpha1/cstor_volume.go [hunk 13]
```diff
@@ -1036,7 +1068,118 @@ spec:
...
+  name: cstor-volume-resize-resizevolume-default
+spec:
+  meta: |
+    id: resizecstorvolume
+    kind: Command
+  post: |
+    {{- $cmd := resize cstor volume -}}
+    {{- $cmd = $cmd | withoption "ip" .TaskResult.resizelistsvc.clusterIP -}}
+    {{- $cmd = $cmd | withoption "volname" .Volume.owner -}}
+    {{- $cmd = $cmd | withoption "capacity" .Volume.capacity -}}
```
Comment (payes) [inline pkg/install/v1alpha1/cstor_volume.go]: colons should be there in all the above cmds,
```
    {{- $cmd := $cmd | withoption "ip" .TaskResult.resizelistsvc.clusterIP -}}
    {{- $cmd := $cmd | withoption "volname" .Volume.owner -}}
    {{- $cmd := $cmd | withoption "capacity" .Volume.capacity -}}
```

File: pkg/volume/volume.go [hunk 14]
```diff
@@ -207,6 +248,108 @@ func (v *Operation) Create() (*v1alpha1.CASVolume, error) {
...
+
+	oldSize := RoundUpToBytes(pv.Spec.Capacity[core_v1.ResourceStorage])
+	glog.V(4).Infof("old capacity: '%d' and requested capacity: '%d' in Bytes", oldSize, newSize)
+
+	// If volume size is already greater than or equal to requested size return
+	if uint64(oldSize) >= newSize {
+		return nil, fmt.Errorf("Volume is already greater than or equal to requested volume size")
+	}
+
+	if string(v.volume.Namespace) != string(pv.Spec.ClaimRef.Namespace) {
```
Comment (payes) [inline pkg/volume/volume.go]: Namespace should be fetched and updated from the ClaimRef. CSI will not be able to pass the namespace for resize. Only PV name will be received.

**Comment by mittachaitu** [comment]
Handled in different PRs:
1) cstor-volume-mgmt: https://github.com/openebs/maya/pull/1361
2) maya-apiserver: https://github.com/openebs/maya/pull/1363
3) CSI: https://github.com/openebs/csi/pull/13

**Pull request closed without merging**